packages/firestore/src/util/bundle.ts

```diff
@@ -0,0 +1,215 @@
+        controller.close();
+      }
+    }
+  });
+}
+
+/**
+ * A class representing a bundle.
+ *
+ * Takes a bundle stream or buffer, and presents abstractions to read bundled
+ * elements out of the underlying content.
+ */
+export class Bundle {
+  // Cached bundle metadata.
+  private metadata?: BundleMetadata | null;
+  // The reader instance of the given ReadableStream.
+  private reader: ReadableStreamDefaultReader;
+  // Internal buffer to hold bundle content, accumulating incomplete element content.
+  private buffer: Uint8Array = new Uint8Array();
+  private textDecoder = new TextDecoder('utf-8');
+
+  constructor(
+    private bundleStream:
+      | ReadableStream<Uint8Array | ArrayBuffer>
+      | Uint8Array
+      | ArrayBuffer
+  ) {
+    if (
+      bundleStream instanceof Uint8Array ||
+      bundleStream instanceof ArrayBuffer
+    ) {
+      this.bundleStream = toReadableStream(bundleStream);
+    }
+    this.reader = (this.bundleStream as ReadableStream).getReader();
+  }
+
+  /**
+   * Returns the metadata of the bundle.
+   */
+  async getMetadata(): Promise<BundleMetadata> {
+    if (!this.metadata) {
+      const result = await this.nextElement();
+      if (result === null || result instanceof SizedBundleElement) {
+        throw new Error(`The first element is not metadata, it is ${result}`);
+      }
+      this.metadata = (result as BundleElement).metadata;
+    }
+
+    return this.metadata!;
+  }
+
+  /**
+   * Asynchronously iterate through all bundle elements (except bundle metadata).
+   */
+  async *elements(): AsyncIterableIterator<SizedBundleElement> {
+    let element = await this.nextElement();
+    while (element !== null) {
+      if (element instanceof SizedBundleElement) {
+        yield element;
+      } else {
+        this.metadata = element.metadata;
+      }
+      element = await this.nextElement();
+    }
+  }
+
+  // Reads from the head of internal buffer, and pulling more data from underlying stream if a complete element
+  // cannot be found, until an element(including the prefixed length and the JSON string) is found.
+  //
+  // Once a complete element is read, it is dropped from internal buffer.
+  //
+  // Returns either the bundled element, or null if we have reached the end of the stream.
+  private async nextElement(): Promise<
+    BundleElement | SizedBundleElement | null
+  > {
+    const lengthBuffer = await this.readLength();
+    if (lengthBuffer === null) {
+      return null;
+    }
+
+    const lengthString = this.textDecoder.decode(lengthBuffer);
+    const length = parseInt(lengthString, 10);
+    if (isNaN(length)) {
+      throw new Error(`length string (${lengthString}) is not valid number`);
+    }
+
+    const jsonString = await this.readJsonString(lengthBuffer.length, length);
+    // Update the internal buffer to drop the read length and json string.
+    this.buffer = this.buffer.slice(lengthBuffer.length + length);
+
+    if (!this.metadata) {
```

Comment:
Metadata being the first element of a stream is a feature, IMO. I cannot think of a use case where metadata being in the middle of a stream is useful, and I think it's better to reject bundles without metadata upfront, than try to cope with it.

I am willing to add it if someone actually find this useful, and request for that.

Reply:
How about we compromise on this one:
- The logic to parse the bundles will be agnostic of the position in the stream. I suspect that this will make the code much easier to understand.
- We throw/reject the user Promise if the bundle stream is invalid.  This should be secondary check that has a clear error (something like: "First element in bundle is not a metadata element")